### PR TITLE
ARROW-12480: [Java][Dataset] FileSystemDataset: Support reading from a directory

### PR DESCRIPTION
**Pull request by zhztheplayer**

### CODE REVIEW

File: cpp/src/arrow/dataset/discovery.h
Comment (zhztheplayer) [inline cpp/src/arrow/dataset/discovery.h:236]: Removed this method from the PR since some of the factory options might vary according to URI, e.g. We may want to enable partitioning when reading folders and disable partitioning when reading a single file. But before URI is resolved we don't even know it's a file or a folder. We may add URI based API back once we have a better design on it.

**Comment by zhztheplayer** [comment]
@emkornfield - Are you able to have a review for this patch? Just a fundamental functionality to make Java dataset API able to read from files in folder.

**Comment by zhztheplayer** [comment]
Ping @kou @liyafan82 @kszucs - If we have other active Java code committers?

**Comment by emkornfield** [comment]
Sorry, not sure when I will be able to take a look (possibility of this weekend) but I'm generally behind on reviews.

**Comment by kou** [comment]
I'm not familiar with Java. @kisizk Can you review this?

**Comment by kou** [comment]
I'm not familiar with Java. @kiszk Can you review this?

**Comment by kiszk** [comment]
Sure, I will do a review.

File: java/dataset/src/test/java/org/apache/arrow/dataset/file/TestFileSystemDataset.java
Comment (kiszk) [inline java/dataset/src/test/java/org/apache/arrow/dataset/file/TestFileSystemDataset.java:160]: Can we use `try (... datum = ...) { ... }` at line 153? So, we can remove this line.

File: java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java
Comment (kiszk) [inline java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java:53]: nit: Do we need this format change?

File: java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java
Comment (kiszk) [inline java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java:51]: I think that this change wants to get a unique name for a short period.

How about using `System.currentTimeMillis()`?

**Comment by kiszk** [comment]
Looks good to me with a few comments.

I would like to get another review, for example, by @emkornfield or @liyafan82

File: java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java
Comment (liyafan82) [inline java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java:46]: Do we need to provide a seed for the object?

File: cpp/src/arrow/dataset/discovery.cc
Comment (liyafan82) [inline cpp/src/arrow/dataset/discovery.cc:219]: nit: it seems we sometimes use `arrow::fs` and sometimes use `fs::`. Is there any special reason for that?

File: java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java
Comment (zhztheplayer) [reply java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java:46]: Can you share some thoughts doing that? Thanks.

File: java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java
Comment (zhztheplayer) [reply java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java:51]: Doesn't milli cause collisions? `ParquetWriteSupport` is instantiated rapidly during ut execution.

By the way any special reason to use time rather than random? The files can be dropped after finishing ut. Users should not be aware of the files.

File: java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java
Comment (zhztheplayer) [reply java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java:53]: It's just a trivial change to keep code readable.

File: java/dataset/src/test/java/org/apache/arrow/dataset/file/TestFileSystemDataset.java
Comment (zhztheplayer) [reply java/dataset/src/test/java/org/apache/arrow/dataset/file/TestFileSystemDataset.java:160]: It seems that `datum` is of type `List` that is not `AutoClosable`.

File: cpp/src/arrow/dataset/discovery.cc
Comment (zhztheplayer) [reply cpp/src/arrow/dataset/discovery.cc:219]: Not special reason for that. `fs::` looks more reasonable since it's used from other places in this file. Changed to `fs::`.

File: java/dataset/src/test/java/org/apache/arrow/dataset/file/TestFileSystemDataset.java
Comment (liyafan82) [inline java/dataset/src/test/java/org/apache/arrow/dataset/file/TestFileSystemDataset.java:354]: It seems the batch is closed here, and in the caller. So it is closed twice?

File: cpp/src/arrow/dataset/discovery.cc
Comment (liyafan82) [reply cpp/src/arrow/dataset/discovery.cc:219]: I see. Thanks.

File: java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java
Comment (liyafan82) [reply java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java:46]: I am not sure if it is a problem:
If we do not use a seed, the file name for each run is identical. So the result of the previous run may be overwritten by the next run.

File: java/dataset/src/test/java/org/apache/arrow/dataset/file/TestFileSystemDataset.java
Comment (zhztheplayer) [reply java/dataset/src/test/java/org/apache/arrow/dataset/file/TestFileSystemDataset.java:354]: removed

File: java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java
Comment (zhztheplayer) [reply java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java:46]: > If we do not use a seed, the file name for each run is identical

I am not sure I get you point here correctly, but did you mean "If we do not use a seed, the file name for each run is **not** identical"?

Comment (liyafan82) [reply java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java:46]: I mean if we do not use a seed, it amounts to always setting 0 as the seed. So each time we run the program, we get the same random numbers (they are just pseudo-random), and the same file names.

File: java/dataset/src/test/java/org/apache/arrow/dataset/file/TestFileSystemDataset.java
Comment (liyafan82) [reply java/dataset/src/test/java/org/apache/arrow/dataset/file/TestFileSystemDataset.java:354]: sorry, I do not see it removed?

Comment (zhztheplayer) [reply java/dataset/src/test/java/org/apache/arrow/dataset/file/TestFileSystemDataset.java]: Oops. It should be removed now.

File: java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java
Comment (zhztheplayer) [reply java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java:46]: > it amounts to always setting 0 as the seed

I don't think it's normal to have a fixed seed in Random by default. Are you using a regular version of JVM?

Comment (liyafan82) [reply java/dataset/src/test/java/org/apache/arrow/dataset/ParquetWriteSupport.java:46]: After investigating the source code, I find you are right. The default constructor uses a seed based on the current time.

File: java/dataset/src/test/java/org/apache/arrow/dataset/file/TestFileSystemDataset.java
Comment (liyafan82) [reply java/dataset/src/test/java/org/apache/arrow/dataset/file/TestFileSystemDataset.java:160]: Maybe place this in a finally block?

**Review by kiszk — APPROVED** [review]

**Comment by kszucs** [comment]
Thanks everyone, merging!

**Pull request closed without merging**
